refactor: introduce single-separator split helper (boost::split replacement)

theStack opened the pull request:
This PR adds a simple string split helper `SplitString` that takes use of the spanparsing `Split` function that was first introduced in #13697 (commit fe8a7dcd78cfeedc9a7c705e91384f793822912b). This enables to replace most calls to `boost::split`, in the cases where only a single separator character is used. Note that while previous attempts to replace `boost::split` were controversial (e.g. #13751), this one has a trivial implementation: it merely uses an internal helper (that is unit tested and in regular use with output descriptiors) and converts its result from spans to strings. As a drawback though, not all `boost::split` instances can be tackled.

As a possible optimization, one could return a vector of `std::string_view`s (available since C++17) instead of strings, to avoid copies. This would need more carefulness on the caller sites though, to avoid potential lifetime issues, and it's probably not worth it, considering that none of the places where strings are split are really performance-critical.

theStack committed:
refactor: introduce single-separator split helper `SplitString`

This helper uses spanparsing::Split internally and enables to replace
all calls to boost::split where only a single separator is passed.

Co-authored-by: Martin Ankerl <Martin.Ankerl@gmail.com>
Co-authored-by: MarcoFalke <falke.marco@gmail.com>

MarcoFalke commented on src/util/string.h:
```diff
@@ -15,6 +16,15 @@
+{
+    std::vector<Span<const char>> span_strs = spanparsing::Split(str, sep);
+    std::vector<std::string> result;
+    for (const auto& span_str : span_strs)
```
nit: missing `{`. Also, could move the impl to the cpp file?

```suggestion
    for (const auto& span_str : span_strs) {
```

theStack replied:
Thanks for the review, done (was surprised to see that the string.cpp is actually empty on master).

maflcko reviewed (commented):
Seems ok to use the already existing split function.

theStack commented:
Force-pushed with changes as suggested by MarcoFalke (https://github.com/bitcoin/bitcoin/pull/22953#discussion_r706614179).

Pinging @practicalswift as someone who has always shown interest in potential boost replacements and string processing helpers, with lots of contributions in that area. Maybe you feel like taking a look :)

MarcoFalke commented on src/util/string.cpp:
```diff
@@ -3,3 +3,14 @@
 #include <util/string.h>
+#include <util/spanparsing.h>
+
+[[nodiscard]] std::vector<std::string> SplitString(const std::string& str, char sep)
```
Is the annotation needed in the cpp file as well? At least lock annotations are only supposed to go to the declaration.

theStack replied:
Good point. Quickly `grep`ing the `[[nodiscard]]` annotation in our codebase shows that there are barely any occurences in .cpp files (only for static functions). Removed it.

kiminuo committed:
test: add unit tests for `SplitString` helper

kiminuo commented:
I would still add a unit test even though there are tests for the internal helper. Maybe it's a little bit over the top so feel free to ignore my suggestion.

(Something like this possibly: https://github.com/kiminuo/bitcoin/commit/ac6dd5a921407c3957a7c15a6a5a82af826decb6)

MarcoFalke commented:
Any reason the tests aren't fixed up as well? 

```
src/test/fuzz/script_assets_test_minimizer.cpp:    boost::algorithm::split(words, str, boost::algorithm::is_any_of(","));
```

Also, if you add a unit test, might as extend the `string` fuzz target.

kiminuo commented:
There non-tests occurrences too:

```
src\rpc\server.cpp:
  410          std::vector<std::string> vargNames;
  411:         boost::algorithm::split(vargNames, argNamePattern, boost::algorithm::is_any_of("|"));
  412          auto fr = argsIn.end();

src\test\transaction_tests.cpp:
  72      std::vector<std::string> words;
  73:     boost::algorithm::split(words, strFlags, boost::algorithm::is_any_of(","));
  74  

src\test\fuzz\script_assets_test_minimizer.cpp:
  133      std::vector<std::string> words;
  134:     boost::algorithm::split(words, str, boost::algorithm::is_any_of(","));
  135  
```

If I'm not missing something.

theStack commented:
> I would still add a unit test even though there are tests for the internal helper. Maybe it's a little bit over the top so feel free to ignore my suggestion.

I agree that it's a good idea to add a unit test.

> 
> (Something like this possibly: [kiminuo@ac6dd5a](https://github.com/kiminuo/bitcoin/commit/ac6dd5a921407c3957a7c15a6a5a82af826decb6))

Thanks! As quick feedback, can you change the commit subject to be more verbose (e.g. "test: add unit tests for SplitString helper") and put the test into `src/test/util_tests.cpp`, instead of creating a new file? I'm happy to either add your commit to this PR, or review your PR if you decide to open your own, based on this one.

> Any reason the tests aren't fixed up as well?
> ...

> There non-tests occurrences too:
> ...

Oh, I only `grep`ed for `boost::split`, not being aware that there is also `boost::algorithm::split`, which seems to be a synonym. Force-pushed with changes to tackle also the remaining occurences. Also found that there were some unnecessary boost includes in `src/torcontrol.cpp` that I forgot to remote in the previous push.

kiminuo commented:
> > (Something like this possibly: [kiminuo@ac6dd5a](https://github.com/kiminuo/bitcoin/commit/ac6dd5a921407c3957a7c15a6a5a82af826decb6))
> 
> Thanks! As quick feedback, can you change the commit subject to be more verbose (e.g. "test: add unit tests for SplitString helper") and put the test into `src/test/util_tests.cpp`, instead of creating a new file? I'm happy to either add your commit to this PR, or review your PR if you decide to open your own, based on this one.

Modified here: https://github.com/kiminuo/bitcoin/commit/4516727e02feb1338285068541e869c36d8b1368. Anyway, feel free to modify the commit in whatever way you like :)

MarcoFalke commented:
The fuzz test I mentioned: (might not compile)

```diff
diff --git a/src/test/fuzz/string.cpp b/src/test/fuzz/string.cpp
index 0c1b45b86c..e5a09c0013 100644
--- a/src/test/fuzz/string.cpp
+++ b/src/test/fuzz/string.cpp
@@ -127,6 +127,7 @@ FUZZ_TARGET(string)
         int64_t amount_out;
         (void)ParseFixedPoint(random_string_1, fuzzed_data_provider.ConsumeIntegralInRange<int>(0, 1024), &amount_out);
     }
+    (void)SplitString(random_string_1, fuzzed_data_provider.ConsumeIntegral<char>());
     {
         (void)Untranslated(random_string_1);
         const bilingual_str bs1{random_string_1, random_string_2};

Commit:
fuzz: add `SplitString` fuzz target

theStack commented:
> The fuzz test I mentioned: (might not compile)

Oh, I didn't see that in your previous comment. It compiles, I added another commit 👌

theStack commented on src/util/string.cpp:
```diff
@@ -3,3 +3,14 @@
+    std::vector<Span<const char>> span_strs = spanparsing::Split(str, sep);
+    std::vector<std::string> result;
+    for (const auto& span_str : span_strs) {
+        result.push_back(std::string(span_str.begin(), span_str.end()));
```
Good idea, thanks! Done.

theStack commented on src/util/string.cpp:
```diff
@@ -3,3 +3,14 @@
 #include <util/string.h>
+#include <util/spanparsing.h>
+
+std::vector<std::string> SplitString(const std::string& str, char sep)
```
Seems like a nice way to solve this! Unfortunately, I fear my C++ template skills are currently too lousy to come up with this. Will give it a try on the weekend. That said, I'm also happy to close this PR for another one with a better solution :)

martinus replied:
I think moving the code into the header and changing it into something like that should work:

```cpp
template <typename T = Span<const char>>
std::vector<T> Split(const Span<const char>& sp, char sep)
{
    std::vector<T> ret;
    auto it = sp.begin();
    auto start = it;
    while (it != sp.end()) {
        if (*it == sep) {
            ret.emplace_back(start, it);
            start = it + 1;
        }
        ++it;
    }
    ret.emplace_back(start, it);
    return ret;
}
```

So the code is practically identical, only the return take is now a template argument. Note that I'm typing this on my phone because I don't have access to a computer the next few days so this probably won't compile :)

MarcoFalke replied:
If you decide to do that, it would be possible to define an alias like so:

```cpp
constexpr auto SplitString = &spanparsing::Split<std::string>;

theStack replied:
@martinus: Thanks, I will try that out in a bit!
@MarcoFalke: That's a good idea with the alias, then I only need to change the implementation and keep the rest as it is now.

fanquake commented:
Concept ACK - this is low-overhead for dropping more Boost dependence. Surprised we can't also nuke anything from `test/lint/lint-includes.sh`

laanwj commented:
Concept ACK. I don't think this boost use is very bad (it's a header-only library) but if we can replace it with functionality we already have internally, who not.

> Concept ACK - this is low-overhead for dropping more Boost dependence. Surprised we can't also nuke anything from test/lint/lint-includes.sh

Well, `split` is still used in some places (`httprpc.cpp` for example), this only replaces the single-separator case. It will take some more work to get rid of all `boost/algorithm/string.hpp` use.

theStack commented:
Force-pushed the alternative variant as suggested by @martinus, changing `spanparsing::Split` to a templated function and defining an alias `SplitString` as suggested by @MarcoFalke (see https://github.com/bitcoin/bitcoin/pull/22953#discussion_r708485720 and https://github.com/bitcoin/bitcoin/pull/22953#discussion_r712016183).

MarcoFalke commented on src/test/util_tests.cpp:
```diff
@@ -2078,6 +2078,55 @@ BOOST_AUTO_TEST_CASE(test_spanparsing)
+    {
+        std::vector<std::string> result = SplitString("AAA", 'a');
+        BOOST_CHECK_EQUAL(result.size(), 1);
+        BOOST_CHECK_EQUAL(result[0], "AAA");
```
weird issue on macos:

```
test/util_tests.cpp:2087: error: in "util_tests/test_SplitString": check result[0] == "" has failed [ != ]
test/util_tests.cpp:2095: error: in "util_tests/test_SplitString": check result[1] == "" has failed [ != ]
test/util_tests.cpp:2104: error: in "util_tests/test_SplitString": check result[2] == "" has failed [ != ]
test/util_tests.cpp:2111: error: in "util_tests/test_SplitString": check result[0] == "abc" has failed [abc != abc]
test/util_tests.cpp:2119: error: in "util_tests/test_SplitString": check result[1] == "b" has failed [b != b]
test/util_tests.cpp:2126: error: in "util_tests/test_SplitString": check result[0] == "AAA" has failed [AAA != AAA]

martinus replied:
I think the issue here is that the conversion of the string literals to `Span<const char>` works so that the `\0` is part of the span. So for `""` the size of the span is actually 1 with content "\0", while for `std::string` the size would be 0.

theStack replied:
Ah, good point, that also explains why only the last element of each result vector failed the test. Thanks, fixed.

martinus commented on src/util/string.h:
```diff
@@ -15,6 +16,8 @@
 #include <string>
 #include <vector>
 
+constexpr auto SplitString = &spanparsing::Split<std::string>;
```
I think it would be safer to use a more restrictive interface for `SplitString`, because it's not really intuitive how SplitString behaves with string literals. Something like

```cpp
inline std::vector<std::string> SplitString(std::string_view str, char sep) {
    return spanparsing::Split<std::string>(str, sep);
}
```

MarcoFalke commented on src/test/util_tests.cpp:
```diff
@@ -2078,6 +2078,55 @@ BOOST_AUTO_TEST_CASE(test_spanparsing)
+
+    // Case-sensitivity of the separator.
+    {
+        std::vector<std::string> result = SplitString(std::string("AAA"), 'a');
```
It seems dangerous to offer a function that silently does the wrong thing when passed the "wrong" type

theStack replied:
Agree. This is fixed now by using an explicit interface instead of an alias, as suggested by martinus in https://github.com/bitcoin/bitcoin/pull/22953#discussion_r712700924.

maflcko reviewed (dismissed):
Not a fan of forcing devs to remember to construct a std::string manually each time before calling the function.

martinus commented:
Windows build error seems unrelated, in `wallet_address_types.py`:
```
OSError: [WinError 10048] Only one usage of each socket address (protocol/network address/port) is normally permitted
```

hebasto commented:
See #18548

On Thu, 23 Sep 2021 at 08:54, Martin Leitner-Ankerl <
***@***.***> wrote:

> Windows build error seems unrelated, in wallet_address_types.py:
>
> OSError: [WinError 10048] Only one usage of each socket address (protocol/network address/port) is normally permitted
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/22953#issuecomment-925523479>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH3PXPQDTK3VEUUU4R2KEY3UDK6JVANCNFSM5D246WWQ>
> .
>
-- 

--
Hennadii Stepanov

theStack commented:
Rebased on master to (hopefully) get rid of the Windows CI failure.

practicalswift commented:
Concept ACK

theStack commented:
Rebased on master.

theStack commented:
Rebased on master again.

kiminuo commented:
@theStack Is there a bug or does the PR need another rebase to pass all tests?

theStack commented:
@kiminuo: Thanks for noticing, indeed a rebaster on master was needed since #17631 (commit ef7c8228fd5cf45526518ae2bd5ebdd483e65525) introduced two other single-separator uses of `boost::split` in `./src/rest.cpp`. Done!

theStack commented:
Rebased on master.

theStack commented:
Rebased on master again (there was a conflict in src/rest.cpp after #24098 has been merged).

fanquake commented:
@martinus want to take another look here?

martinus commented:
Code review ACK a62e84438d27ee6213219fe2c233e58814fcbb5d. Ran all tests. I also like that with `boost::split` it was not obvious that the resulting container was cleared, and with `SplitString` API that's obvious.

After that PR there is only one usage of `boost::split` left in `core_read.cpp`, that would be a nice followup PR :slightly_smiling_face:
